runtime/lua/vim/_editor.lua

```diff
@@ -715,6 +715,107 @@ function vim._on_key(buf, typed_buf)
+---@private
+---@param s string
+---@return integer
+local function utf_ptr2char(s)
+  local b1, b2, b3, b4 = s:byte(1, 4)
+
+  if b1 < 0x80 then
+    -- 1-byte sequence (ASCII character)
+    return b1
+  elseif b1 < 0xE0 then
+    -- 2-byte sequence
+    return ((b1 % 0x20) * 0x40) + (b2 % 0x40)
+  elseif b1 < 0xF0 then
+    -- 3-byte sequence
+    return ((b1 % 0x10) * 0x1000) + ((b2 % 0x40) * 0x40) + (b3 % 0x40)
+  else
+    -- 4-byte sequence
+    return ((b1 % 0x08) * 0x40000) + ((b2 % 0x40) * 0x1000) + ((b3 % 0x40) * 0x40) + (b4 % 0x40)
+  end
+end
+
+--- Convert UTF-32 or UTF-16 {index} to byte index.
+--- {opts} can be a table containing two properties:i
+--- * an "encoding" property of "utf-8", "utf-16" or "utf-32"
+--- * an "error" property of true or false, if error is false
+--- then then an out of range index will return byte length
+--- insert of an error.
+---
+--- {opts} can also be supplied as a boolean, true for utf-16 and false for utf-32.
+--- If {opts} is not supplied, it defaults to false (UTF-32). Returns the byte index.
+---
+--- Invalid UTF-8 and NUL is treated like in |vim.str_utfindex()|.
+--- An {index} in the middle of a UTF-16 sequence is rounded upwards to
+--- the end of that sequence.
+---@param s string
+---@param index integer
+---@param opts? boolean|nil| { encoding: "utf-8"|"utf-16"|"utf-32", error?: boolean }
+---@return integer
+function vim.str_byteindex(s, index, opts)
+  local len = #s
+  if index == 0 then
+    return 0
+  end
+
+  if index > len then
+    return len
+  end
+
+  opts = opts or { encoding = 'utf-32', error = true }
+  if type(opts) == 'boolean' then
+    opts = opts and { encoding = 'utf-16', error = true } or { encoding = 'utf-32', error = true }
+  end
+  local encoding = opts.encoding or 'utf-32'
+  local valid_encodings = { ['utf-8'] = true, ['utf-16'] = true, ['utf-32'] = true }
+  if not valid_encodings[encoding] then
+    error('Invalid encoding: ' .. encoding)
+  end
+
+  if opts.encoding == 'utf-8' then
+    return index
+  end
+
+  local byte_i = vim.str_utf_pos(s:gsub('%z', ' '))
+  local utf16_ptr = 0
+  local utf16_char = 0
+  local utf32_ptr = 0
+  local utf32_char = 0
+  -- remove NUL from the string  (UTF-16 and UTF32 both use 4 bytes for NUL)
+  local strlen = vim.fn.strchars(s:gsub('%z', '0000'))
+  local j = 0
+  while j <= strlen do
+    j = j + 1
+    local char_len = vim.str_utf_end(s, byte_i[j])
+    local char = s:sub(byte_i[j], byte_i[j] + char_len) or ''
+    local c = utf_ptr2char(char)
+    if c > 0XFFFF then
+      utf16_char = utf16_char + 2
+    else
+      utf16_char = utf16_char + 1
+    end
+    utf32_char = utf32_char + 1
+
+    utf32_ptr = utf32_ptr + char_len + 1
+    utf16_ptr = utf16_ptr + char_len + 1
+
+    if opts.encoding == 'utf-16' and utf16_char >= index then
+      return utf16_ptr
+    end
+
+    if opts.encoding == 'utf-32' and utf32_char >= index then
+      return utf32_ptr
+    end
+  end
```

Comment:
```suggestion
  -- Prepare string by removing NUL characters
  local prepared_string = s:gsub('%z', ' ')
  local strlen = vim.fn.strchars(prepared_string)
  local utf16_ptr, utf16_char = 0, 0
  local utf32_ptr, utf32_char = 0, 0
  local j = 0

  -- Traverse the string and calculate pointers for UTF-16 and UTF-32
  while j <= strlen do
    j = j + 1
    local char_len = vim.str_utf_end(s, j)
    local c = utf_ptr2char(s:sub(j, j + char_len))

    if c > 0xFFFF then
      utf16_char = utf16_char + 2
    else
      utf16_char = utf16_char + 1
    end
    utf32_char = utf32_char + 1

    if encoding == 'utf-16' and utf16_char >= index then
      return utf16_ptr + char_len
    elseif encoding == 'utf-32' and utf32_char >= index then
      return utf32_ptr + char_len
    end

    utf32_ptr = utf32_ptr + char_len + 1
    utf16_ptr = utf16_ptr + char_len + 1
  end
```

Reply:
I have taken some aspects from this comment to clean the loop up, but I think I need the str_pos() for edge cases around split bytes

Reply:
I have taken a bit of a merged approach, as we did need to track the byte index too for substringing when the byte is split.

I have pushed a separate commit for now, will squash when you have taken a look @glepnir

Reply:
most LGTM just one thing shouldn't `if s:find('%z') then` do sub ? sub will have memory allocation to generate a new string.

Reply:
i have added a conditional check, so the sub is only done if a NUL is found

Reply:
LGTM now.